src/BotPlutusInterface/Balance.hs

```diff
@@ -74,62 +78,69 @@ import BotPlutusInterface.Collateral (removeCollateralFromMap)
-    BalanceTxCollateral
-  | -- | `BalanceTxWithScripts` denotes that we are balancing the `Tx` that uses scripts.
-    BalanceTxWithScripts
-  | -- | `BalanceTxWithoutScripts` denotes that we are balancing the `Tx` that doesn't uses any scripts.
-    BalanceTxWithoutScripts
-  deriving stock (Eq)
-
--- `balanceTxIO` checks the type of the current `UnbalancedTx` and then calls `balanceTxIO'`.
+data TxWithScript
+data TxWithoutScript
+
+data BalanceTxConstraint a where
+  TxWithScript :: BalanceTxConstraint TxWithScript
+  TxWithoutScript :: BalanceTxConstraint TxWithoutScript
+  TxWithSeparateChange :: BalanceTxConstraint a
+
+instance Eq (BalanceTxConstraint a) where
+  TxWithScript == TxWithScript = True
+  TxWithoutScript == TxWithoutScript = True
+  TxWithSeparateChange == TxWithSeparateChange = True
+  _ == _ = False
+
+class KnownBalanceConstraint (a :: Type) where
+  knownConstraint :: proxy a -> BalanceTxConstraint a
+
+instance KnownBalanceConstraint TxWithScript where
+  knownConstraint _ = TxWithScript
+
+instance KnownBalanceConstraint TxWithoutScript where
+  knownConstraint _ = TxWithoutScript
+
+{- | Collect necessary tx inputs and collaterals, add minimum lovelace values and balance non ada
+ assets. `balanceTxIO` assumes that the `Tx` we are balancing is of type `TxWithoutScript`.
+ If you want to add custom constraints while balancing the `Tx` like, saying that `Tx` contains
+ scripts refer to `balanceTxIO'`.
+-}
 balanceTxIO ::
   forall (w :: Type) (effs :: [Type -> Type]).
   (Member (PABEffect w) effs) =>
   PABConfig ->
   PubKeyHash ->
   UnbalancedTx ->
   Eff effs (Either Text Tx)
-balanceTxIO pabConf ownPkh unbalancedTx@(UnbalancedTx tx' _ _ _)
-  | validCollateralTx tx' = balanceTxIO' @w pabConf ownPkh unbalancedTx BalanceTxCollateral
-  | txUsesScripts tx' = balanceTxIO' @w pabConf ownPkh unbalancedTx BalanceTxWithScripts
-  | otherwise = balanceTxIO' @w pabConf ownPkh unbalancedTx BalanceTxWithoutScripts
-  where
-    validCollateralTx :: Tx -> Bool
-    validCollateralTx tx
-      | [out] <- txOutputs tx
-        , Map.size (txData tx) == 0
-        , txMint tx == mempty
-        , txOutValue out == collateralValue pabConf
-        , txMintScripts tx == mempty
-        , txCollateral tx == mempty
-        , not (txUsesScripts tx) =
-        True
-      | otherwise =
-        False
+balanceTxIO = balanceTxIO' @w [TxWithoutScript]
 
-{- | Collect necessary tx inputs and collaterals, add minimum lovelace values and balance non ada
-     assets
+{- | This is just a more flexible version of `balanceTxIO` which let us specify the `BalanceTxConstraint`(s)
+ -   for the `Tx` that we are balancing.
 -}
 balanceTxIO' ::
-  forall (w :: Type) (effs :: [Type -> Type]).
-  (Member (PABEffect w) effs) =>
+  forall (w :: Type) (effs :: [Type -> Type]) (a :: Type).
+  (Member (PABEffect w) effs, KnownBalanceConstraint a) =>
+  [BalanceTxConstraint a] ->
```

Comment:
Might this be better as a set, passing in duplicate constraints doesn't make much sense.

Reply:
fixed